### DISCUSSION TITLE
Signature mismatches in Server

### DISCUSSION THREAD

**Discussion by marcus-j-davies** [zwave-js/ZWaveJS.NET#32] [General]
@raman325

**Controller Firmware Update** VS **Node Firmware Update**

I have started a discussion here, as its not allowed on the repo.
I am going full throttle at bringing the .NET lib up to date with the latest Schema (https://github.com/zwave-js/ZWaveJS.NET/pull/31) - I'm not far from completing it.

But in doing so I have spotted a signature, that doesn't seem necessary.

firmwareUpdateOTW - only needs a blob
https://github.com/zwave-js/node-zwave-js/blob/11d23a6796fe5b0324a9e6566364c34fceb1cca3/packages/zwave-js/src/lib/controller/Controller.ts#L6718

But you seem to require a lot more

https://github.com/zwave-js/zwave-js-server/blob/16b8fb359720132e77799ac3b174ca829c480e7c/src/lib/controller/incoming_message.ts#L273

https://github.com/zwave-js/zwave-js-server/blob/16b8fb359720132e77799ac3b174ca829c480e7c/src/lib/controller/message_handler.ts#L313

Your interface for this command seems to be a match with updating a ZWave Node, and not the Controller it's self?
I don't actually know what will happen if you pass a Controllers FW to the Driver method of `extractFirmware`

I don't want to start creating  issues left right and centre, without first passing you my findings, hence what I find, I'll first discuss here?

EDIT:
I _think_ you have noted this on the read me.
Admittedly  - I may not be aware of the reasons for some of these Quirks - so forgive me if there are reasons 😅

**Comment by marcus-j-davies** [comment]
@raman325

Wrong property name requested in the `update` interface for `node.updateFirmware`

https://github.com/zwave-js/zwave-js-server/blob/16b8fb359720132e77799ac3b174ca829c480e7c/src/lib/node/message_handler.ts#L81

is expecting the update interface to contain a `file` property (Buffer), but the docs in the driver have `data` instead

https://zwave-js.github.io/node-zwave-js/#/api/node?id=updatefirmware

You do some internal processing here - which escapes the issue (I think), but all it takes is for a user to refer to documents on the driver, and use the correct name to cause an error.

My suggestion is to instead.

 - Correct the property name
 - Expose the method `extractFirmware` to Socket clients
 
 This will remove the need for the internal hackery, and align closer to the driver docs

**Comment by raman325** [comment]
@marcus-j-davies generally speaking all of these don't match the spec as expected - I just didn't see a reason why a user should have to send multiple websocket commands back and forth to achieve the same thing they could if they just sent it in one. With that being said, open to moving closer back to the raw API. Will have to take a look deeper before I comment more
